File: src/utils/shapeGeometry.ts
import type { ShapeType } from '../types/canvas';

/** Regular polygon points centered in a width×height box (local coords). */
export function regularPolygonPoints(
  sides: number,
  width: number,
  height: number,
): number[] {
  const cx = width / 2;
  const cy = height / 2;
  const rx = width / 2;
  const ry = height / 2;
  const pts: number[] = [];
  const start = -Math.PI / 2;
  for (let i = 0; i < sides; i++) {
    const a = start + (i * 2 * Math.PI) / sides;
    pts.push(cx + rx * Math.cos(a), cy + ry * Math.sin(a));
  }
  return pts;
}

export function diamondPoints(width: number, height: number): number[] {
  return [width / 2, 0, width, height / 2, width / 2, height, 0, height / 2];
}

export function trianglePoints(width: number, height: number): number[] {
  return [width / 2, 0, width, height, 0, height];
}

export function parallelogramPoints(width: number, height: number): number[] {
  const skew = width * 0.22;
  return [skew, 0, width, 0, width - skew, height, 0, height];
}

export function starPointsLocal(
  width: number,
  height: number,
  points = 5,
  innerRatio = 0.45,
): number[] {
  const cx = width / 2;
  const cy = height / 2;
  const rx = width / 2;
  const ry = height / 2;
  const pts: number[] = [];
  const start = -Math.PI / 2;
  for (let i = 0; i < points * 2; i++) {
    const a = start + (i * Math.PI) / points;
    const r = i % 2 === 0 ? 1 : innerRatio;
    pts.push(cx + rx * r * Math.cos(a), cy + ry * r * Math.sin(a));
  }
  return pts;
}

/** Speech-bubble-like callout: rounded body + triangular tail. */
export function calloutPoints(width: number, height: number): number[] {
  const bodyH = height * 0.72;
  const r = Math.min(width, bodyH) * 0.12;
  const tailX = width * 0.28;
  const tailW = width * 0.16;
  // Approximate with polygon (body rectangle + tail)
  return [
    r,
    0,
    width - r,
    0,
    width,
    r,
    width,
    bodyH - r,
    width - r,
    bodyH,
    tailX + tailW,
    bodyH,
    tailX + tailW * 0.35,
    height,
    tailX,
    bodyH,
    r,
    bodyH,
    0,
    bodyH - r,
    0,
    r,
  ];
}

export function shapePointsFor(
  shapeType: ShapeType,
  width: number,
  height: number,
  opts?: { starPoints?: number; starInnerRatio?: number },
): number[] | null {
  const w = Math.max(width, 1);
  const h = Math.max(height, 1);
  switch (shapeType) {
    case 'triangle':
      return trianglePoints(w, h);
    case 'diamond':
      return diamondPoints(w, h);
    case 'pentagon':
      return regularPolygonPoints(5, w, h);
    case 'hexagon':
      return regularPolygonPoints(6, w, h);
    case 'star':
      return starPointsLocal(w, h, opts?.starPoints ?? 5, opts?.starInnerRatio ?? 0.45);
    case 'parallelogram':
      return parallelogramPoints(w, h);
    case 'callout':
      return calloutPoints(w, h);
    default:
      return null;
  }
}

File: lib/core/canvas/geometry/shape_geometry.dart
import 'dart:math' as math;

import '../models/canvas_element.dart';
import 'point.dart';
import 'rect.dart';

/// Build a closed polygon path (list of points) for a shape in world bounds.
List<Point> shapePolygon(ShapeKind kind, Rect2 bounds, {int starPoints = 5, double starInnerRatio = 0.45}) {
  final cx = bounds.center.x;
  final cy = bounds.center.y;
  final w = bounds.width;
  final h = bounds.height;
  final left = bounds.left;
  final top = bounds.top;
  final right = bounds.right;
  final bottom = bounds.bottom;

  switch (kind) {
    case ShapeKind.rectangle:
    case ShapeKind.roundedRect:
      return [
        Point(left, top),
        Point(right, top),
        Point(right, bottom),
        Point(left, bottom),
      ];
    case ShapeKind.ellipse:
      // Approximate for hit-testing polygons; ellipse uses dedicated hit.
      const n = 32;
      return [
        for (var i = 0; i < n; i++)
          Point(
            cx + (w / 2) * math.cos(2 * math.pi * i / n),
            cy + (h / 2) * math.sin(2 * math.pi * i / n),
          ),
      ];
    case ShapeKind.triangle:
      return [
        Point(cx, top),
        Point(right, bottom),
        Point(left, bottom),
      ];
    case ShapeKind.diamond:
      return [
        Point(cx, top),
        Point(right, cy),
        Point(cx, bottom),
        Point(left, cy),
      ];
    case ShapeKind.pentagon:
      return _regularPolygon(cx, cy, w / 2, h / 2, 5, -math.pi / 2);
    case ShapeKind.hexagon:
      return _regularPolygon(cx, cy, w / 2, h / 2, 6, 0);
    case ShapeKind.star:
      return _star(cx, cy, w / 2, h / 2, starPoints, starInnerRatio);
    case ShapeKind.parallelogram:
      final skew = w * 0.25;
      return [
        Point(left + skew, top),
        Point(right, top),
        Point(right - skew, bottom),
        Point(left, bottom),
      ];
    case ShapeKind.callout:
      final tailH = h * 0.22;
      final bodyBottom = bottom - tailH;
      return [
        Point(left, top),
        Point(right, top),
        Point(right, bodyBottom),
        Point(left + w * 0.45, bodyBottom),
        Point(left + w * 0.28, bottom),
        Point(left + w * 0.35, bodyBottom),
        Point(left, bodyBottom),
      ];
  }
}

List<Point> _regularPolygon(
  double cx,
  double cy,
  double rx,
  double ry,
  int sides,
  double startAngle,
) {
  return [
    for (var i = 0; i < sides; i++)
      Point(
        cx + rx * math.cos(startAngle + 2 * math.pi * i / sides),
        cy + ry * math.sin(startAngle + 2 * math.pi * i / sides),
      ),
  ];
}

List<Point> _star(
  double cx,
  double cy,
  double rx,
  double ry,
  int points,
  double innerRatio,
) {
  final result = <Point>[];
  final n = points * 2;
  for (var i = 0; i < n; i++) {
    final angle = -math.pi / 2 + math.pi * i / points;
    final outer = i.isEven;
    final rScale = outer ? 1.0 : innerRatio;
    result.add(Point(
      cx + rx * rScale * math.cos(angle),
      cy + ry * rScale * math.sin(angle),
    ));
  }
  return result;
}

bool pointInPolygon(Point point, List<Point> polygon) {
  if (polygon.length < 3) return false;
  var inside = false;
  for (var i = 0, j = polygon.length - 1; i < polygon.length; j = i++) {
    final xi = polygon[i].x;
    final yi = polygon[i].y;
    final xj = polygon[j].x;
    final yj = polygon[j].y;
    final intersect = ((yi > point.y) != (yj > point.y)) &&
        (point.x <
            (xj - xi) * (point.y - yi) / ((yj - yi) == 0 ? 1e-9 : (yj - yi)) +
                xi);
    if (intersect) inside = !inside;
  }
  return inside;
}

double distanceToSegment(Point p, Point a, Point b) {
  final dx = b.x - a.x;
  final dy = b.y - a.y;
  final lengthSq = dx * dx + dy * dy;
  if (lengthSq == 0) {
    final ddx = p.x - a.x;
    final ddy = p.y - a.y;
    return math.sqrt(ddx * ddx + ddy * ddy);
  }
  var t = ((p.x - a.x) * dx + (p.y - a.y) * dy) / lengthSq;
  t = t.clamp(0.0, 1.0);
  final projX = a.x + t * dx;
  final projY = a.y + t * dy;
  final ddx = p.x - projX;
  final ddy = p.y - projY;
  return math.sqrt(ddx * ddx + ddy * ddy);
}

double distanceToPolyline(Point p, List<double> points) {
  if (points.length < 4) return double.infinity;
  var min = double.infinity;
  for (var i = 0; i + 3 < points.length; i += 2) {
    final d = distanceToSegment(
      p,
      Point(points[i], points[i + 1]),
      Point(points[i + 2], points[i + 3]),
    );
    if (d < min) min = d;
  }
  return min;
}

Rect2 normalizeRect(double x1, double y1, double x2, double y2) {
  return Rect2.fromPoints(Point(x1, y1), Point(x2, y2));
}

/// Constrain to square / circle when shift is held.
Rect2 constrainProportional(Rect2 rect, Point start, Point current) {
  final size = math.max(rect.width, rect.height);
  final dx = current.x >= start.x ? size : -size;
  final dy = current.y >= start.y ? size : -size;
  return Rect2.fromPoints(start, Point(start.x + dx, start.y + dy));
}
